src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/RazorProjectChangePublisher.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal abstract class RazorProjectChangePublisher : ProjectSnapshotChangeTrigger
+    {
+        private ProjectSnapshotManagerBase _projectSnapshotManager;
+        protected readonly ConcurrentDictionary<string, string> _publishFilePathMappings;
+
+        public RazorProjectChangePublisher()
+        {
+            _publishFilePathMappings = new ConcurrentDictionary<string, string>(FilePathComparer.Instance);
+        }
+
+        public override void Initialize(ProjectSnapshotManagerBase projectManager)
+        {
+            _projectSnapshotManager = projectManager;
+            _projectSnapshotManager.Changed += ProjectSnapshotManager_Changed;
+        }
+
+        internal abstract bool IsLSPEditorAvailable(string projectFilePath);
```

Comment:
This feels really odd to have on this class. Why not on the feature detector?

Reply:
I'm not a fan this arrangement either, but it was the best way I could find to abstract these out so-as not to break our project chain and testing. Basically if the `LSPEditorFeatureDetector` gets called from the tests they break because those services haven't been populated. I suppose I could just make it internal virtual on `DefaultRazorProjectChangePublisher` instead of abstract here since its arguably an implementation detail.

Reply:
I'd think of this as an impl detail but more specifically i'd imagine the impl would call through to the feature detector.